pkg/container-utils/containerd/containerd.go

```diff
@@ -15,31 +15,275 @@
 package containerd
 
 import (
+	"context"
+	"fmt"
 	"time"
 
-	criclient "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/cri"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/errdefs"
+	"github.com/sirupsen/logrus"
+
 	runtimeclient "github.com/inspektor-gadget/inspektor-gadget/pkg/container-utils/runtime-client"
 )
 
 const (
 	DefaultTimeout = 2 * time.Second
+
+	LabelK8sContainerName         = "io.kubernetes.container.name"
+	LabelK8sContainerdKind        = "io.cri-containerd.kind"
+	LabelK8sContainerdKindSandbox = "sandbox"
 )
 
+var log = logrus.WithField("container-client", "containerd")
+
 type ContainerdClient struct {
-	criclient.CRIClient
+	client *containerd.Client
 }
 
 func NewContainerdClient(socketPath string) (runtimeclient.ContainerRuntimeClient, error) {
 	if socketPath == "" {
 		socketPath = runtimeclient.ContainerdDefaultSocketPath
 	}
 
-	criClient, err := criclient.NewCRIClient(runtimeclient.ContainerdName, socketPath, DefaultTimeout)
+	client, err := containerd.New(socketPath,
```

Comment:
I am getting this error on a node w/o containerd:

```
Error: initializing manager: failed to dial "/run/containerd/containerd.sock": context deadline exceeded: connection error: desc = "transport: error while dialing: dial unix:///run/containerd/containerd.sock: timeout"
```
This means I won't be able to use `ig` on a node w/o containerd without specifying `runtimes` flag.

Reply:
@blanquicet do you have any thoughts on this? I think `ig` should keep working even if one of the runtime isn't available. What do you say?

Reply:
Oh no. It totally should work without `containerd`. That is not the intended behavior. Thanks for catching this

Reply:
Now I remember what happens:
All other clients have a lazy initialization. When creating the `docker` or `crio` client nothing happens even if there is no socket. Only on the first call they try to connect to the socket and then an error happens. See the following output:
```
WARN[0000] Runtime enricher (docker): couldn't get current containers: listing containers with options {Size:false All:true Latest:false Since: Before: Limit:0 Filters:{fields:map[]}}: Cannot connect to the Docker daemon at unix:///run/docker.sock. Is the docker daemon running? 
WARN[0002] Runtime enricher (containerd): failed to initialize container runtime: failed to dial "/run/containerd/containerd.sock": context deadline exceeded: connection error: desc = "transport: error while dialing: dial unix:///run/containerd/containerd.sock: timeout" 
```
With `docker` the error appears when getting the current containers. Here we only print the error. With the `containerd` client the error instantly happens when initializing the client. Here we return the error.

So should i change that to only printing a warning?

Reply:
> So should i change that to only printing a warning?

I believe so like `podman` also has lazy client but I am trying to understand if that is how runtime clients are supposed to be used.

Reply:
Yep, it is. From the beginning of `ig`, we decided that even if the client fails to initialize, `ig` shouldn't fail and it should continue trying to use that client to enrich future containers because maybe the runtime was not temporarily working when we launched `ig`.

Reply:
> It should continue trying to use that client to enrich future containers because maybe the runtime was not temporarily working when we launched `ig`.

Currently it is now working this way thought for any container enricher. I tested it with a docker container and renamed the `docker.sock` to something else. Started `ig` and then renamed `docker.sock` back to its original name. No events are showing up

Reply:
You mean it working not working for future containers ? For me it works fine for docker:

```
$ sudo systemctl stop docker && sudo systemctl stop docker.socket
$ sudo ig version
v0.17.0
$ sudo ig list-containers -r docker -w
WARN[0000] Runtime enricher (docker): couldn't get current containers: listing containers with options {Size:false All:true Latest:false Since: Before: Limit:0 Filters:{fields:map[]}}: Cannot connect to the Docker daemon at unix:///run/docker.sock. Is the docker daemon running? 
TIMESTAMP                      EVENT      RUNTIME    ID
# events from future
2023-06-23T12:05:56+02:00      CREATED    docker     59818fa298d961fb47c21bdf27202c50813dd6f46a8679e… angry_kirch8                                                                                                 
2023-06-23T12:05:58+02:00      DELETED    docker     59818fa298d961fb47c21bdf27202c50813dd6f46a8679e… angry_kirch8  

# goto another terminal
$ sudo systemctl start docker && sudo systemctl start docker.socket
$ docker run -it busybox

Reply:
> You mean it working not working for future containers ? For me it works fine for docker:

Yes i meant not working for future containers. But apparently it works for some cases (like you tested it with a stopped docker service.
Whereas I tested it with renaming `/run/docker.sock` to `/run/docker.sock.foo`. Starting `ig` and while it runs renaming `/run/docker.sock.foo` back to /run/docker.sock`

Reply:
Thanks to @mqasimsarfraz the behavior is now the same as before and all other clients.
When creating our own `grpc` connection and giving it to the `containerd` client the connection seems to be lazily initialized.